### ISSUE TITLE
diff2html@2.0.0-beta9 not working on client side.

### ISSUE THREAD

**Issue by codingtwinky** [rtfpessoa/diff2html#65]
Looking at the code borwserifying and importing nunjucks at client side will not be helpful either.  

```
TypeError: nunjucks.configure is not a function
    at http://127.0.0.1:8333/js/ungit.js:2895:30
    at Object.<anonymous> (http://127.0.0.1:8333/js/ungit.js:2910:3)
    at Object.require.24../diff-parser.js (http://127.0.0.1:8333/js/ungit.js:2912:4)
    at s (http://127.0.0.1:8333/js/ungit.js:1:262)
    at http://127.0.0.1:8333/js/ungit.js:1:313
    at http://127.0.0.1:8333/js/ungit.js:2710:23
    at Object.require.23../diff-parser.js (http://127.0.0.1:8333/js/ungit.js:2873:3)
    at s (http://127.0.0.1:8333/js/ungit.js:1:262)
    at http://127.0.0.1:8333/js/ungit.js:1:313
    at http://127.0.0.1:8333/js/ungit.js:2675:27
    at Object.require.22../line-by-line-printer.js (http://127.0.0.1:8333/js/ungit.js:2693:3)
    at s (http://127.0.0.1:8333/js/ungit.js:1:262)
    at http://127.0.0.1:8333/js/ungit.js:1:313
    at http://127.0.0.1:8333/js/ungit.js:18316:21
    at Object.<anonymous> (http://127.0.0.1:8333/js/ungit.js:18415:3)
    at Object.require.diff2html../diff-parser.js (http://127.0.0.1:8333/js/ungit.js:18417:4)
```

**Comment by rtfpessoa** [comment]
`nunjucks` was added as a dependency for version `2.0.0`. Can you confirm it was downloaded?

Are you requiring the code from `src` or `dist`? You should use `src`. `dist` is for the browser.

**Comment by codingtwinky** [comment]
I was using disted version.  Maybe because I'm doing browserify on diff2html and nunjucks being browserified twice??

I will try to use src later and let you know

**Comment by rtfpessoa** [comment]
I am currently using webpack for the dist version. And I am passing the parameter saying it is for WEB.

I am not sure if this creates any kinds of problems.

When running webpack I do this transformations:

```
    new webpack.NormalModuleReplacementPlugin(
      /(nunjucks)$/,
      'nunjucks/index'
    ),

    new webpack.NormalModuleReplacementPlugin(
      /(precompile|nodes|lexer|parser|transformer|compiler|loaders)$/,
      'node-libs-browser/mock/empty'
    ),

    new webpack.DefinePlugin({
      'process.env': {
        IS_BROWSER: true
      }
    })
```

And I also precompile the templates to `dist/diff2html-templates.js`

If you give me some tips about browserify maybe I can also do a dist with it (not sure if it makes sense)

**Comment by orivej** [comment]
The problem is that for browserify nunjuncks in `var nunjucks = require('nunjucks');` is provided by `nunjucks/browser/nunjucks.js`, browserify expects nunjucks to set external `module` and `exports`, but `nunjucks/browser/nunjucks.js` does not touch them, it just assigns its module to `var nunjucks`, and the module remains unconfigured. (I do not know yet how to solve this.)

**Comment by orivej** [comment]
This is an open issue with nunjucks: https://github.com/mozilla/nunjucks/issues/507

**Comment by codingtwinky** [comment]
Thanks for the discovery, I be keeping my eyes on that issue and upgrade diff2html accordingly.

**Comment by rtfpessoa** [comment]
@orivej as I said in the previous answer if you use:

```
    new webpack.NormalModuleReplacementPlugin(
      /(nunjucks)$/,
      'nunjucks/index'
    ),

    new webpack.NormalModuleReplacementPlugin(
      /(precompile|nodes|lexer|parser|transformer|compiler|loaders)$/,
      'node-libs-browser/mock/empty'
    ),

    new webpack.DefinePlugin({
      'process.env': {
        IS_BROWSER: true
      }
    })
```

Webpack works just fine, is there any way to replace modules in browserify?

**Comment by orivej** [comment]
I'm using webpack too; I missed that this snippet is something I should try to put into my `webpack.config.js`, rather than just something you use to build a release of diff2html.

I tried to include it as is, and observed that the build fails since the second plugin hides important modules unrelated to nunjucks. Need to figure out how to circumvent this…

**Comment by rtfpessoa** [comment]
@orivej it is a really bad hack. I am wondering if nunjucks is going to give a lot of headaches.

**Comment by pbu88** [comment]
Sorry for being a bit away these last weeks. Is a pity that nunjucks have this issue which such a hacky workaround. We hoped nunjucks would make it easier to have better formatted and easier to maintain and override html but if is not working for us we might want to looks for alternative solutions. Do you guys see a value in using it at all? Other solutions were proposed, like using React components for the html.

I would like to know what's your opinion on this one?

**Comment by rtfpessoa** [comment]
This seems to be really bad. @orivej @codingtwinky any suggestions about other options to move the html from the JS code?

I am starting to regret nunjucks. It nows seems even more complicated. Since we are only in beta for this we can roll back to other option.

**Comment by codingtwinky** [comment]
I don't know much about nunjucks so I can't do fair comparisons, but if you are looking for html templating frameworks isn't [handlebars](http://handlebarsjs.com/) or [mustache](https://github.com/janl/mustache.js/) are the standard goto solutions?

**Comment by orivej** [comment]
I will have to look into the way diff2html uses nunjucks to give an informed judgement.

**Comment by rtfpessoa** [comment]
@codingtwinky thanks for the input. I will also take a look at thoose options a make some tests when I have some time.

**Comment by pbu88** [comment]
@rtfpessoa also, ECMAScript 6 allows multiline strings with backticks ( \` ). I've seen a lot of people using ECMAScript 6 and transpiling it with Babel to make it compatible. Is just something to consider maybe. I like React but I think it would be better to keep the project dependencies on the low and React is a huge dependency but that's just my opinion.

**Comment by rtfpessoa** [comment]
Now I also think react is too much for the purpose.

I took a look at mustache and handlebars and both look good. Mustache seems simpler but you cannot have loops etc. For that reason Handlebars seems a good alternative since it is simple and has precompiled templates. I will do some tests and see if I find any other alternative.

In the meanwhile let me know if you have any other good suggestions.

**Comment by rtfpessoa** [comment]
I have a couple more alternatives here (for no special order):

* [hogan.js](http://twitter.github.io/hogan.js/) from Twitter
* [dust.js](http://www.dustjs.com/) from Linkedin
* [pug.js](https://github.com/pugjs/pug) influenced by Haml (previously named Jade)
* [JsRender](https://github.com/BorisMoore/jsrender)
* [marko.js](http://markojs.com/github/)

From what I have seen [hogan.js](http://twitter.github.io/hogan.js/) seems to be a good alternative to handlebars since it also implements mustache and is more lightweight.

Any thoughts @codingtwinky @orivej @pbu88 ?

**Comment by codingtwinky** [comment]
I have never done client side or web server programming in my life professionally, I'm just a backend data infrastructure grunt who have terrible habit of poking javascript open sources.  But I did have lots fun with Jade and did like it alot as you can import and export templates with in the page really easily.  I can't believe I have forgotten about it.

But my vote shouldn't count as much as others and other libs may have better features.

**Comment by rtfpessoa** [comment]
I want something lightweight with just the minimum set of features.

I was trying with hogan.js but I am having some problems using the templates on the browser and on the node side at the same time.

**Comment by rtfpessoa** [comment]
I think #69 fixes this issue, version 2.0.0-beta11, can you check?

**Comment by orivej** [comment]
I had to add `var Hogan = require('hogan.js');` to `diff2html-templates.js` to build a project using diff2html with webpack. `try/catch` around `require` in HoganJsUtils prolonged the time to debug.

**Comment by rtfpessoa** [comment]
@orivej can you check it on `2.0.0-beta12` it should fix the problem.

**Comment by orivej** [comment]
As of `2.0.0-beta12`, diff2html works as is.

With `2.0.0-beta11`, templates did not render and there were no exceptions in the browser console, so I began debugging by examining if webpack produced sensible output: it was not obvious since it wrapped the contents of `diff2html-templates.js` with `(function(global) {`…`}.call(exports, (function() { return this; }())))`, and it was not until I removed `try/catch` from `HoganJsUtils` that I saw the cause of null rendering.

**Comment by rtfpessoa** [comment]
@orivej just released `2.0.0-beta13`, now I am considering that I will always compile the templates before release.

FYI, I have also been adding lots of more tests to the parser and the generators.
Specially the parser is now more stable and I fixed several corner cases.

**Comment by rtfpessoa** [comment]
I am considering this fixed with PR #69
